<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Изучение-данных" data-toc-modified-id="Изучение-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Изучение данных</a></span></li><li><span><a href="#Предобратка-данных" data-toc-modified-id="Предобратка-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Предобратка данных</a></span><ul class="toc-item"><li><span><a href="#Дубликаты" data-toc-modified-id="Дубликаты-1.2.1"><span class="toc-item-num">1.2.1&nbsp;&nbsp;</span>Дубликаты</a></span></li><li><span><a href="#Пропуски" data-toc-modified-id="Пропуски-1.2.2"><span class="toc-item-num">1.2.2&nbsp;&nbsp;</span>Пропуски</a></span></li></ul></li><li><span><a href="#Подготовка-признаков" data-toc-modified-id="Подготовка-признаков-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Подготовка признаков</a></span><ul class="toc-item"><li><span><a href="#Удаление-ненужных-признаков" data-toc-modified-id="Удаление-ненужных-признаков-1.3.1"><span class="toc-item-num">1.3.1&nbsp;&nbsp;</span>Удаление ненужных признаков</a></span></li><li><span><a href="#Прямое-кодирование" data-toc-modified-id="Прямое-кодирование-1.3.2"><span class="toc-item-num">1.3.2&nbsp;&nbsp;</span>Прямое кодирование</a></span></li><li><span><a href="#Разделение-на-обучающую,-валидационную-и-тестовую-выборки" data-toc-modified-id="Разделение-на-обучающую,-валидационную-и-тестовую-выборки-1.3.3"><span class="toc-item-num">1.3.3&nbsp;&nbsp;</span>Разделение на обучающую, валидационную и тестовую выборки</a></span></li><li><span><a href="#Масштабирование-признаков" data-toc-modified-id="Масштабирование-признаков-1.3.4"><span class="toc-item-num">1.3.4&nbsp;&nbsp;</span>Масштабирование признаков</a></span></li></ul></li></ul></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Случайный лес</a></span></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Взвешивание-классов" data-toc-modified-id="Взвешивание-классов-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Взвешивание классов</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-3.1.3"><span class="toc-item-num">3.1.3&nbsp;&nbsp;</span>Случайный лес</a></span></li></ul></li><li><span><a href="#Увеличение-выборки" data-toc-modified-id="Увеличение-выборки-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Увеличение выборки</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Случайный лес</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span></li></ul></div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Постройте модель с предельно большим значением *F1*-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте *F1*-меру на тестовой выборке самостоятельно.

Дополнительно измеряйте *AUC-ROC*, сравнивайте её значение с *F1*-мерой.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

# Инструкция по выполнению проекта

1. Загрузите и подготовьте данные. Поясните порядок действий.
2. Исследуйте баланс классов, обучите модель без учёта дисбаланса. Кратко опишите выводы.
3. Улучшите качество модели, учитывая дисбаланс классов. Обучите разные модели и найдите лучшую. Кратко опишите выводы.
4. Проведите финальное тестирование.

# Признаки

RowNumber — индекс строки в данных  
CustomerId — уникальный идентификатор клиента  
Surname — фамилия  
CreditScore — кредитный рейтинг  
Geography — страна проживания  
Gender — пол  
Age — возраст  
Tenure — сколько лет человек является клиентом банка  
Balance — баланс на счёте  
NumOfProducts — количество продуктов банка, используемых клиентом  
HasCrCard — наличие кредитной карты  
IsActiveMember — активность клиента  
EstimatedSalary — предполагаемая зарплата  

## Подготовка данных

### Изучение данных

In [1]:
from sklearn.linear_model import LogisticRegression
import warnings
from sklearn.utils import shuffle
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import f1_score, confusion_matrix, recall_score, precision_score
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt
import pandas as pd
pd.options.display.max_columns = None
pd.options.mode.chained_assignment = None
warnings.filterwarnings("ignore")

In [2]:
df = pd.read_csv('/datasets/Churn.csv')
df = df.drop('RowNumber', axis=1)
df.columns = df.columns.str.lower()

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 13 columns):
customerid         10000 non-null int64
surname            10000 non-null object
creditscore        10000 non-null int64
geography          10000 non-null object
gender             10000 non-null object
age                10000 non-null int64
tenure             9091 non-null float64
balance            10000 non-null float64
numofproducts      10000 non-null int64
hascrcard          10000 non-null int64
isactivemember     10000 non-null int64
estimatedsalary    10000 non-null float64
exited             10000 non-null int64
dtypes: float64(3), int64(7), object(3)
memory usage: 1015.8+ KB


Обращает на себя внимание отсутствие данных о длительности существования клиента (tenure) по 909 клиентам.
Для столбцов geography, gender можно осуществить прямое кодирование.
Стобцам hascard, isactivemember, exited можно предать булев тип (но не нужно).

In [4]:
display(df.head())

customerid   surname  creditscore geography  gender  age  tenure  \
0    15634602  Hargrave          619    France  Female   42     2.0   
1    15647311      Hill          608     Spain  Female   41     1.0   
2    15619304      Onio          502    France  Female   42     8.0   
3    15701354      Boni          699    France  Female   39     1.0   
4    15737888  Mitchell          850     Spain  Female   43     2.0   

     balance  numofproducts  hascrcard  isactivemember  estimatedsalary  \
0       0.00              1          1               1        101348.88   
1   83807.86              1          0               1        112542.58   
2  159660.80              3          1               0        113931.57   
3       0.00              2          0               0         93826.63   
4  125510.82              1          1               1         79084.10   

   exited  
0       1  
1       0  
2       1  
3       0  
4       0

In [5]:
pd.set_option('display.float_format', lambda x: '%.4f' % x)
display(df.describe())

customerid  creditscore        age    tenure     balance  \
count    10000.0000   10000.0000 10000.0000 9091.0000  10000.0000   
mean  15690940.5694     650.5288    38.9218    4.9977  76485.8893   
std      71936.1861      96.6533    10.4878    2.8947  62397.4052   
min   15565701.0000     350.0000    18.0000    0.0000      0.0000   
25%   15628528.2500     584.0000    32.0000    2.0000      0.0000   
50%   15690738.0000     652.0000    37.0000    5.0000  97198.5400   
75%   15753233.7500     718.0000    44.0000    7.0000 127644.2400   
max   15815690.0000     850.0000    92.0000   10.0000 250898.0900   

       numofproducts  hascrcard  isactivemember  estimatedsalary     exited  
count     10000.0000 10000.0000      10000.0000       10000.0000 10000.0000  
mean          1.5302     0.7055          0.5151      100090.2399     0.2037  
std           0.5817     0.4558          0.4998       57510.4928     0.4028  
min           1.0000     0.0000          0.0000          11.5800     0.0000  
25%           1.0000     0.0000          0.0000       51002.1100     0.0000  
50%           1.0000     1.0000          1.0000      100193.9150     0.0000  
75%           2.0000     1.0000          1.0000      149388.2475     0.0000  
max           4.0000     1.0000          1.0000      199992.4800     1.0000

Никаких явных искажений не наблюдается.  
20% выборки перестали пользоваться услугами банка. Это, конечно, много. 
Это и отражает баланс классов.

In [6]:
display(df.corr())

customerid  creditscore     age  tenure  balance  \
customerid           1.0000       0.0053  0.0095 -0.0214  -0.0124   
creditscore          0.0053       1.0000 -0.0040 -0.0001   0.0063   
age                  0.0095      -0.0040  1.0000 -0.0131   0.0283   
tenure              -0.0214      -0.0001 -0.0131  1.0000  -0.0079   
balance             -0.0124       0.0063  0.0283 -0.0079   1.0000   
numofproducts        0.0170       0.0122 -0.0307  0.0120  -0.3042   
hascrcard           -0.0140      -0.0055 -0.0117  0.0272  -0.0149   
isactivemember       0.0017       0.0257  0.0855 -0.0322  -0.0101   
estimatedsalary      0.0153      -0.0014 -0.0072  0.0105   0.0128   
exited              -0.0062      -0.0271  0.2853 -0.0168   0.1185   

                 numofproducts  hascrcard  isactivemember  estimatedsalary  \
customerid              0.0170    -0.0140          0.0017           0.0153   
creditscore             0.0122    -0.0055          0.0257          -0.0014   
age                    -0.0307    -0.0117          0.0855          -0.0072   
tenure                  0.0120     0.0272         -0.0322           0.0105   
balance                -0.3042    -0.0149         -0.0101           0.0128   
numofproducts           1.0000     0.0032          0.0096           0.0142   
hascrcard               0.0032     1.0000         -0.0119          -0.0099   
isactivemember          0.0096    -0.0119          1.0000          -0.0114   
estimatedsalary         0.0142    -0.0099         -0.0114           1.0000   
exited                 -0.0478    -0.0071         -0.1561           0.0121   

                 exited  
customerid      -0.0062  
creditscore     -0.0271  
age              0.2853  
tenure          -0.0168  
balance          0.1185  
numofproducts   -0.0478  
hascrcard       -0.0071  
isactivemember  -0.1561  
estimatedsalary  0.0121  
exited           1.0000

Предварительно можно отметить, что самая большая корреляция с оттоком клиентов у следующих признаков (в порядке убывания):  
- возраст
- активность клиента (связь обратнопропорциональная)
- баланс

### Предобратка данных

#### Дубликаты

In [7]:
df.duplicated().sum()

0

#### Пропуски

In [8]:
display(df.query('tenure.isna()').head())

customerid    surname  creditscore geography  gender  age  tenure  \
30    15589475    Azikiwe          591     Spain  Female   39     nan   
48    15766205        Yin          550   Germany    Male   38     nan   
51    15768193  Trevisani          585   Germany    Male   36     nan   
53    15702298   Parkhill          655   Germany    Male   41     nan   
60    15651280     Hunter          742   Germany    Male   35     nan   

       balance  numofproducts  hascrcard  isactivemember  estimatedsalary  \
30      0.0000              3          1               0      140469.3800   
48 103391.3800              1          0               1       90878.1300   
51 146050.9700              2          0               0       86424.5700   
53 125561.9700              1          0               0      164040.9400   
60 136857.0000              1          0               0       84509.5700   

    exited  
30       1  
48       0  
51       0  
53       1  
60       0

In [9]:
report = df.query('tenure.isna()')['exited'].value_counts().to_frame()
report.columns = ['counts']
report['% of total'] = (report['counts'] * 100 /
                        report['counts'].sum()).round(2)
display(report)

counts  % of total
0     726     79.8700
1     183     20.1300

Целевой признак оттока exits распределен в пропусках почти так же, как в датафрейме в целом. Это позволяет нам избавиться от данных без изменения в соотношении целевого признака.

In [10]:
df = df.dropna()
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9091 entries, 0 to 9998
Data columns (total 13 columns):
customerid         9091 non-null int64
surname            9091 non-null object
creditscore        9091 non-null int64
geography          9091 non-null object
gender             9091 non-null object
age                9091 non-null int64
tenure             9091 non-null float64
balance            9091 non-null float64
numofproducts      9091 non-null int64
hascrcard          9091 non-null int64
isactivemember     9091 non-null int64
estimatedsalary    9091 non-null float64
exited             9091 non-null int64
dtypes: float64(3), int64(7), object(3)
memory usage: 994.3+ KB


### Подготовка признаков

#### Удаление ненужных признаков

In [11]:
df = df.drop(['surname', 'customerid'], axis=1)

#### Прямое кодирование

In [12]:
ohe_list = ['geography', 'gender']
df_ohe = pd.get_dummies(df, columns=ohe_list, drop_first=True)
df_ohe.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9091 entries, 0 to 9998
Data columns (total 12 columns):
creditscore          9091 non-null int64
age                  9091 non-null int64
tenure               9091 non-null float64
balance              9091 non-null float64
numofproducts        9091 non-null int64
hascrcard            9091 non-null int64
isactivemember       9091 non-null int64
estimatedsalary      9091 non-null float64
exited               9091 non-null int64
geography_Germany    9091 non-null uint8
geography_Spain      9091 non-null uint8
gender_Male          9091 non-null uint8
dtypes: float64(3), int64(6), uint8(3)
memory usage: 736.9 KB


#### Разделение на обучающую, валидационную и тестовую выборки

In [13]:
target = df_ohe['exited']
features = df_ohe.drop('exited', axis=1)
features_general, features_test, target_general, target_test = train_test_split(
    features, target, test_size=0.20, random_state=12345)
features_train, features_valid, target_train, target_valid = train_test_split(
    features_general, target_general, test_size=0.25, random_state=12345)


print(features.shape)
print(features_test.shape)
print(features_general.shape)
print(features_train.shape)
print(features_valid.shape)

(9091, 11)
(1819, 11)
(7272, 11)
(5454, 11)
(1818, 11)


#### Масштабирование признаков

In [14]:
numeric = ['creditscore', 'age', 'tenure',
           'balance', 'numofproducts', 'estimatedsalary']

scaler = StandardScaler()
scaler.fit(features_train[numeric])

features_train[numeric] = scaler.transform(features_train[numeric])
features_general[numeric] = scaler.transform(features_general[numeric])
features_valid[numeric] = scaler.transform(features_valid[numeric])
features_test[numeric] = scaler.transform(features_test[numeric])

## Исследование задачи

In [15]:
display(features_train.head())

creditscore     age  tenure  balance  numofproducts  hascrcard  \
3706      -0.2038  0.4713  0.3523   0.7860         0.7824          0   
6805      -0.3575 -0.3849 -1.3735  -1.2306         0.7824          1   
4449       0.1753 -0.2898 -0.6832  -1.2306         0.7824          1   
598        0.3495  1.7080  0.0072   1.3795        -0.9149          0   
1845       0.9028 -0.2898  1.3878  -1.2306        -0.9149          0   

      isactivemember  estimatedsalary  geography_Germany  geography_Spain  \
3706               0          -0.3572                  0                1   
6805               1          -1.6710                  0                0   
4449               0          -1.1192                  0                0   
598                0          -1.5691                  1                0   
1845               1           1.5438                  0                0   

      gender_Male  
3706            0  
6805            0  
4449            1  
598             0  
1845            1

In [16]:
def build_model(features_train_def, target_train_def, features_valid_def, target_valid_def, type_of_model, depth=20, n_est=60, **kwargs):

    if type_of_model == 'LogisticRegression':
        model = LogisticRegression(
            random_state=12345, solver='liblinear', **kwargs)
        model.fit(features_train_def, target_train_def)
        predicted_valid = model.predict(features_valid_def)
        print("F1 модели:", f1_score(target_valid_def, predicted_valid))
        print("Матрица ошибок модели:")
        display(pd.DataFrame(
            confusion_matrix(target_valid_def, predicted_valid),
            columns=['Pred 0', 'Pred 1'],
            index=['True 0', 'True 1']))
        print("Полнота модели: ", recall_score(
            target_valid_def, predicted_valid))
        print("Точность модели: ", precision_score(
            target_valid_def, predicted_valid))

    if type_of_model == 'DecisionTreeClassifier':
        best_model = None
        best_f1_score = 0
        for depth in range(1, depth):
            model = DecisionTreeClassifier(
                random_state=12345, max_depth=depth, **kwargs)
            model.fit(features_train_def, target_train_def)
            predicted_valid = model.predict(features_valid_def)
            model_f1_score = f1_score(target_valid_def, predicted_valid)
            model_confusion_matrix = confusion_matrix(
                target_valid_def, predicted_valid)
            model_recall_score = recall_score(
                target_valid_def, predicted_valid)
            model_precision_score = precision_score(
                target_valid_def, predicted_valid)
            if model_f1_score > best_f1_score:
                best_model = model
                best_confusion_matrix = model_confusion_matrix
                best_recall_score = model_recall_score
                best_precision_score = model_precision_score
                best_f1_score = model_f1_score
                best_depth = depth
        print("F1 наилучшей модели:", best_f1_score)
        print("Матрица ошибок наилучшей модели:")
        display(pd.DataFrame(best_confusion_matrix, columns=[
                'Pred 0', 'Pred 1'], index=['True 0', 'True 1']))
        print("Полнота наилучшей модели: ", best_recall_score)
        print("Точность наилучшей модели: ", best_precision_score)
        print("Максимальная глубина:", best_depth)

    if type_of_model == 'RandomForestClassifier':
        best_model = None
        best_f1_score = 0
        for est in range(10, n_est, 10):
            for depth in range(1, depth):
                model = RandomForestClassifier(
                    random_state=12345, n_estimators=est, max_depth=depth, **kwargs)
                model.fit(features_train_def, target_train_def)
                predicted_valid = model.predict(features_valid_def)
                model_f1_score = f1_score(target_valid_def, predicted_valid)
                model_confusion_matrix = confusion_matrix(
                    target_valid_def, predicted_valid)
                model_recall_score = recall_score(
                    target_valid_def, predicted_valid)
                model_precision_score = precision_score(
                    target_valid_def, predicted_valid)
                if model_f1_score > best_f1_score:
                    best_model = model
                    best_confusion_matrix = model_confusion_matrix
                    best_recall_score = model_recall_score
                    best_precision_score = model_precision_score
                    best_f1_score = model_f1_score
                    best_depth = depth
                    best_est = est
        print("F1 наилучшей модели:", best_f1_score)
        print("Матрица ошибок наилучшей модели:")
        display(pd.DataFrame(best_confusion_matrix, columns=[
                'Pred 0', 'Pred 1'], index=['True 0', 'True 1']))
        print("Полнота наилучшей модели: ", best_recall_score)
        print("Точность наилучшей модели: ", best_precision_score)
        print("Максимальная глубина:", best_depth)
        print("Кол-во деревьев:", best_est)

### Логистическая регрессия

In [17]:
build_model(features_train, target_train, features_valid,
            target_valid, 'LogisticRegression')

F1 модели: 0.3004115226337448
Матрица ошибок модели:


Pred 0  Pred 1
True 0    1405      47
True 1     293      73

Полнота модели:  0.1994535519125683
Точность модели:  0.6083333333333333


Логистическая регрессия показывает быстрый, но не точный результат

### Дерево решений

In [18]:
build_model(features_train, target_train, features_valid,
            target_valid, 'DecisionTreeClassifier')

F1 наилучшей модели: 0.557427258805513
Матрица ошибок наилучшей модели:


Pred 0  Pred 1
True 0    1347     105
True 1     184     182

Полнота наилучшей модели:  0.4972677595628415
Точность наилучшей модели:  0.6341463414634146
Максимальная глубина: 9


Дерево намного лучше, но все еще не дотягивает до порогового значения F1

### Случайный лес

In [19]:
build_model(features_train, target_train, features_valid,
            target_valid, 'RandomForestClassifier')

F1 наилучшей модели: 0.5570934256055363
Матрица ошибок наилучшей модели:


Pred 0  Pred 1
True 0    1401      51
True 1     205     161

Полнота наилучшей модели:  0.43989071038251365
Точность наилучшей модели:  0.7594339622641509
Максимальная глубина: 14
Кол-во деревьев: 20


Случайный лес показал F1 score не лучше, чем дерево решений. Увеличение точности обернулось снижением полноты.

## Борьба с дисбалансом

### Взвешивание классов

#### Логистическая регрессия

In [20]:
build_model(features_train, target_train, features_valid,
            target_valid, 'LogisticRegression', class_weight='balanced')

F1 модели: 0.49506903353057197
Матрица ошибок модели:


Pred 0  Pred 1
True 0    1055     397
True 1     115     251

Полнота модели:  0.6857923497267759
Точность модели:  0.3873456790123457


Взвешивание классов позволило добиться значительного прогресса в части показателя F1 для логистической регрессии.

#### Дерево решений

In [21]:
build_model(features_train, target_train, features_valid,
            target_valid, 'DecisionTreeClassifier', class_weight='balanced')

F1 наилучшей модели: 0.5555555555555556
Матрица ошибок наилучшей модели:


Pred 0  Pred 1
True 0    1103     349
True 1      91     275

Полнота наилучшей модели:  0.7513661202185792
Точность наилучшей модели:  0.4407051282051282
Максимальная глубина: 7


Взвешивание классов в деревере решений ухудшило результат. Скорее всего, это просто случайность: 0.5570934256055363 против 0.5555555555555556. Изменились и параметры лучшей модели.

#### Случайный лес

In [22]:
build_model(features_train, target_train, features_valid,
            target_valid, 'RandomForestClassifier', class_weight='balanced')

F1 наилучшей модели: 0.595460614152203
Матрица ошибок наилучшей модели:


Pred 0  Pred 1
True 0    1292     160
True 1     143     223

Полнота наилучшей модели:  0.6092896174863388
Точность наилучшей модели:  0.5822454308093995
Максимальная глубина: 9
Кол-во деревьев: 50


0.595460614152203 против 0.5570934256055363. Знаительный прогресс. Также изменились параметры лучшей модели.

### Увеличение выборки

In [23]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)

    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345)

    return features_upsampled, target_upsampled

In [24]:
report = target_train.value_counts().to_frame()
report.columns = ['counts']
report['% of total'] = (report['counts'] * 100 /
                        report['counts'].sum()).round(2)
display(report)

counts  % of total
0    4335     79.4800
1    1119     20.5200

In [25]:
features_upsampled, target_upsampled = upsample(
    features_train, target_train, 4)

#### Логистическая регрессия

In [26]:
build_model(features_upsampled, target_upsampled,
            features_valid, target_valid, 'LogisticRegression')

F1 модели: 0.4956605593056895
Матрица ошибок модели:


Pred 0  Pred 1
True 0    1038     414
True 1     109     257

Полнота модели:  0.7021857923497268
Точность модели:  0.3830104321907601


#### Дерево решений

In [27]:
build_model(features_upsampled,
            target_upsampled,
            features_valid,
            target_valid,
            'DecisionTreeClassifier')

F1 наилучшей модели: 0.5555555555555556
Матрица ошибок наилучшей модели:


Pred 0  Pred 1
True 0    1103     349
True 1      91     275

Полнота наилучшей модели:  0.7513661202185792
Точность наилучшей модели:  0.4407051282051282
Максимальная глубина: 7


#### Случайный лес

In [28]:
build_model(features_upsampled, 
            target_upsampled, 
            features_valid,
            target_valid, 
            'RandomForestClassifier')

F1 наилучшей модели: 0.6055488540410132
Матрица ошибок наилучшей модели:


Pred 0  Pred 1
True 0    1240     212
True 1     115     251

Полнота наилучшей модели:  0.6857923497267759
Точность наилучшей модели:  0.5421166306695464
Максимальная глубина: 9
Кол-во деревьев: 40


### Вывод

Метод увеличения выборки в случайном лесе наиболее успешен.

## Тестирование модели

Используем нашу лучшую модель с заданными праметрами, но дообученной на валидационной выборке.

In [29]:
report = target_general.value_counts().to_frame()
report.columns = ['counts']
report['% of total'] = (report['counts'] * 100 /
                        report['counts'].sum()).round(2)
display(report)

counts  % of total
0    5787     79.5800
1    1485     20.4200

In [30]:
features_general_upsampled, target_general_upsampled = upsample(
    features_general, target_general, 4)

In [31]:
model = RandomForestClassifier(
    random_state=12345, n_estimators=40, max_depth=9)
model.fit(features_general_upsampled, target_general_upsampled)
predicted_test = model.predict(features_test)
print("F1 модели:", f1_score(target_test, predicted_test))
print("Матрица ошибок модели:")
display(pd.DataFrame(
    confusion_matrix(target_test, predicted_test),
    columns=['Pred 0', 'Pred 1'],
    index=['True 0', 'True 1']))
print("Полнота модели: ", recall_score(target_test, predicted_test))
print("Точность модели: ", precision_score(target_test, predicted_test))

F1 модели: 0.6199770378874857
Матрица ошибок модели:


Pred 0  Pred 1
True 0    1218     232
True 1      99     270

Полнота модели:  0.7317073170731707
Точность модели:  0.5378486055776892


Цель достигнута. Модель показывает 0.6199